### ISSUE TITLE
More Articles / Tutorials

### ISSUE THREAD

**Issue by LoopControl** [wagenaartje/neataptic#8]
I'm a relative newbie to Neural networks and have been playing with a few different JS libraries for a couple weeks since they seem to have so much potential.

I recently made a self-driving car that seeks targets on the screen using synaptic.js. I trained it by having it take 3 directional movements (go forward, turn left, turn right) then calculated the distance to the target. Whichever of the 3 actions got the closest, I'd use synaptic's `.propogate()` method to teach it to do that "correct" movement action.

This seems to work OK and after 10s of thousands of iterations, the cars are successfully trained to seek targets.

I was curious however if I could upgrade the cars ability by using neaptatic to 'evolve' a better car without having to tell it what I think the correct movement action to take at every step (I'd like to just have the Neataptic execute one of the 3 random actions then just give the network a score/reward based on distance to the target so the network can figure out what actions to take and evolve better cars).

I repeatedly went over your XOR example but it seems you're actually teaching the network by giving the correct answer (`Methods.Cost.MSE([0], genome.activate([0, 0]))`) which seems closer to what I'm currently doing with Synaptic.js? Is there a way to teach these things via a positive/negative reward system instead of feeding it "correct" answers?

Is this kind of learning possible with this library? Could you give me some pointers on how I could do this kind of teaching with Neataptic or perhaps add some tutorials/articles that explain how to do this kind of thing?

Thanks and sorry this is so long.

**Comment by wagenaartje** [comment]
This is exactly what I want people to do with Neataptic actually :). Your idea is definetily possible, at the moment, im working on something like this (a feedback learning system) for agario. It's working pretty well.

At this moment im on a trip though, but somewhere in 7 days I will post this example, and perhaps ill make a quick versiom for the cars too.

I requires your fitness function to be continues; aka add/remove score until the car is dead or some other feature has been reached.

But somewhere soon ill make some examples for this!

**Comment by LoopControl** [comment]
That's great to hear! Enjoy your trip and I look forward to your insights :)

**Comment by wagenaartje** [comment]
Thanks! Just a quick update: I have been busy with the agar.io neural networks. It seems to be working to a certain extent, I'm tweaking the parameters a little and I hope I can post an article on it by tomorrow.

https://i.gyazo.com/215edb276934c3b19fdaf30547bb63ac.mp4

After that i'll get started on the target seeking cars!

**Comment by LoopControl** [comment]
Looks good, thanks for the update!

**Comment by wagenaartje** [comment]
Indeed, multiple agents are competing against each other in the video.  This speeds up the evolution process, although I think having one agent compete against other already optimized algorithms would also be interesting to see.

Small update:

https://i.gyazo.com/52a69aa57327ffb363a9dd54a0c06e37.mp4

I definitely see some human-like aspects (avoiding big blobs, going after small blobs). The above result is after ~1 hour of running the algorithm. I'll run it for another hour or so and see if there are any improvements (I don't really like the twitching of the blobs).

**Comment by wagenaartje** [comment]
I have made some commits related to this issue:

- [Update README.md](https://github.com/wagenaartje/neataptic/blob/master/README.md) - emphasis on the evolution of neural agents, added Agar.IO example and target seeking cars example (to come!)
- [Article on Agar.IO-AI](https://wagenaartje.github.io/neataptic/articles/agario/) - please do mention if anything important is missing here
- [A repo of the Agar.IO-AI](https://github.com/wagenaartje/agario-ai)

Somewhere later this week i'll get started on target seeking cars.

**Comment by LoopControl** [comment]
Looks great! I just read through the article and will go through the code to see how things are implemented. Thanks a bunch!

**Comment by wagenaartje** [comment]
So I started on the target seeking thingy, I instructed genomes to get as close as possible to a randomly selected other genome in the field. 

Let's just say they seemed to realise that all going to the same location would give them great fitness.

https://i.gyazo.com/fdf65eeaccee5f4595faff8a971fd9e3.mp4 
Evil buggers. :rage:

**Comment by LoopControl** [comment]
Interesting. I've started playing with my own seeking thing based on your agar article and am having some limited success so far.

Maybe making the world wrap around would stop them from going-to/getting-stuck-in corners in your demo?

**Comment by wagenaartje** [comment]
Yes I changed it to wrap around just now, sometimes I notice my population doing fairly good on the seeking.

Youre completely right, I should mention that in the article and the readme of the repo. The reason your code isn't working is because `neat.population` is an array of `Network`s, instead of just one `Network`.

So if you scroll down to the bottom of [population.js](https://github.com/wagenaartje/agario-ai/blob/master/js/population.js), you'll see this code:

```js
ar newPop = [];
for(var i = 0; i < PLAYER_AMOUNT; i++){
  var json = population[i % population.length];
  newPop[i] = neataptic.Network.fromJSON(json);
}
population = newPop;
```

which looks quite complicated, but if you want to simplify it:

```js
var trained = JSON.stringify(neat.population); // Save

for(var i in trained){
  trained[i] = neataptic.Network.fromJSON(trained[i]);
}

neat.population = trained; // Load
```

So basically for every json in the array, convert it to a network. I might have made a type because I didn't test this, but i'll add it to the article soon.

**Comment by LoopControl** [comment]
Thanks again. Got the saving/loading working. I've managed to get some decent seeking behavior now:

https://gfycat.com/ThoseImperfectGarpike

**Comment by wagenaartje** [comment]
Looks good, keep me updated on the progress!

**Comment by LoopControl** [comment]
I ran into similar issues and haven't found a great workaround yet. I start all the seekers at one random location X distance away from a random Target location to make sure they don't spawn close.

I also added the moving/animated target location to make sure seekers are actually following it around and not just there by chance. The problem I'm seeing is that seekers seem to gravitate to the center even when the target moves away because they know the target will eventually choose another random location and have a high chance of coming back near them.

**Comment by wagenaartje** [comment]
So I managed to get something working:

https://gyazo.com/f91f9b7041796d3bb8730dda73474b44

I set the initial positions of all genomes and the target to the same location - they must stay as close to the target as possible.

**Comment by LoopControl** [comment]
Thats a huge improvement!

**Comment by wagenaartje** [comment]
After 10 generations at least one genome has 'figured it out' already, but this is how it looks like after 500+ gens. The only problem now is that the networks get too big, so I'm going to put a constraint on that.

https://i.gyazo.com/69a21d283c0109d521f634eaf64406eb.mp4

I'll upload the code to a repo soon.

**Comment by wagenaartje** [comment]
Last update for some time:

* [Article on Target-seeking AI](https://wagenaartje.github.io/neataptic/articles/targetseeking/)
* [Repo for Target-Seeking AI](https://github.com/wagenaartje/target-seeking-ai)

I also want to say that I have added a `import()` and `export()` to the `neat` class.

I'm going to focus on updating the source code for some time now. So no further examples some time. Once I'm ready to make na example again, I want to create a complex collision avoiding simulation with neural networks.

**Comment by LoopControl** [comment]
Thanks a bunch for these articles, they're super helpful and go through the whole evolution process in Neataptic really well.

And the `import` and `export` are a nice touch as well. Good luck with the core development!
